exercises/tree-building/examples/success-standard/src/TreeBuilding.hs

```diff
@@ -0,0 +1,77 @@
+module TreeBuilding (newTree, Record(..), Tree(..)) where
+
+import Data.List
+import Data.Maybe (isNothing)
+
+type Id = Int
+type Children = [Tree]
+type ParentGrouping = (Maybe Id, [Id])
+
+data Record = Record Id (Maybe Id) deriving (Eq, Show)
+
+data Tree = Leaf Id | Branch Id Children deriving (Eq, Show)
+
+newTree :: [Record] -> Maybe Tree
+newTree records
+  | cycles records      = Nothing
+  | succIdCheck records = build . groupByParent $ records
+  | otherwise           = Nothing
+
+-- Checks for cycling in newTree
+cycles :: [Record] -> Bool
+cycles = any checkRecord
+  where checkRecord r =
+          case r of
+            (Record i Nothing)  -> if' (i /= 0) True False
+            (Record i (Just p)) -> if' (p >= i) True False
+
+build :: [ParentGrouping] -> Maybe Tree
+build [] = Nothing
+build (x:xs)
+  | not . validRoot $ x = Nothing
+  | null xs             = Just (Leaf 0)
+  | otherwise           = Just (Branch 0 (build' xs))
+
+--  Internal build
+build' :: [ParentGrouping] -> [Tree]
+build' xs
+  | length xs == 1 = map Leaf (snd (head xs))
+  | otherwise      = buildWithChildren xs
+  where
+    buildWithChildren (y:ys) =
+      map (\yid ->
+             let children = filter ((== Just yid) . fst) ys
+             in if not (null children)
+                then Branch yid (build' children)
+                else Leaf yid
+          ) (snd y)
```

Comment:
This function is not exhaustive.

I might write it like:

```haskell
build' :: [ParentGrouping] -> [Tree]
build' [(_, yids)] = map Leaf yids
build' ((_, yids):pgs) = map buildWithChildren yids
  where
    buildWithChildren yid =
      let children = [ child | child <- pgs, fst child == Just yid ]
      in if null children
         then Leaf yid
         else Branch yid (build' children)
```

And I might address that it's not exhaustive by concluding that it [shouldn't be empty](http://hackage.haskell.org/package/base/docs/Data-List-NonEmpty.html).